### PR TITLE
Add strawman repository guidelines

### LINKED ISSUE #3521: Open Enclave Repository Guidelines
Currently [OE](https://github.com/openenclave) has a number of repos:
* openenclave: SDK for developing enclaves
* openenclave-build: Reproducible & secure build resources for Open Enclave
* oeedger8r-cpp: An implementation of oeedger8r in C++
* openenclave-mbedtls: Open Enclave copy of mbedtls for inclusion as submodule in SDK.
* openenclave-site: Website for Open Enclave SDK, please file issues at the main repo.
* openenclave-ci
* gperftools: Main gperftools repository (forked from gperftools/gperftools)
* public-talks: Common area for sharing community content
* openenclave-curl: libcurl build files to enable use in Open Enclave applications
* openenclave-template: This is a template repo with a license and code of conduct.

plus two more listed as "private".   However, there is no currently documented process for creating or removing such repos.  Furthermore, some of the existing repos do not use the same license.  For example, while most use MIT, gperftools (being a fork of gperftools/gperftools) uses BSD-3-Clause, and public-talks uses Apache.   

The [Open Enclave Technical Charter](https://github.com/openenclave/openenclave/blob/master/TECHNICAL_CHARTER.md) states that the MIT License is the default, but that:

> The TSC may approve the use of an alternative license or licenses for inbound or outbound contributions on an exception basis. To request an exception, please describe the contribution, the alternative open source license(s), and the justification for using an alternative open source license for the Project. License exceptions must be approved by a two-thirds vote of the entire TSC.

I don't believe the CGC ever explicitly discussed the license issue for public-talks, which just illustrates that we need to better document the process by which repos are approved.   Similarly, forking of other repos for purposes of creating upstream patches (e.g., gperftools) is another area where licenses may vary.

Besides license, other questions for new repos include:
a) should github issues be filed in the new repo, or in the main openenclave repo?
b) should issues in the new repo be discussed in the master openenclave triage meeting, or in some separate process?
c) should one or more of the existing SIGs (e.g., SIG-Architecture, SIG-Testing) oversee relevant aspects of the new repo, or not?
d) will the repo be public or private, and if private, why?

An example of a  writeup of repo guidelines is the [Kubernetes Repository Guidelines](https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md).

### PR DESCRIPTION
**Pull request by dthaler**
Add some proposed guidelines around repositories and creating/deleting them.  The structure of this doc is loosely based off the [Kubernetes Repository Guidelines](https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md) but the content is tailored to OE.

Fixes #3521 

Signed-off-by: Dave Thaler <dthaler@ntdev.microsoft.com>

### CODE REVIEW

**Comment by BRMcLaren** [comment]
Would the act of changing repository ownership to open enclave from a different organization follow the same process as creating repositories?

File: docs/Community/Contributing.md [hunk 1]
```diff
@@ -316,7 +316,9 @@ Copying Files from Other Projects
 
 Open Enclave uses some files from other projects, typically to provide a default
 level of functionality within the enclave where a binary distribution does not
-exist or would be inconvenient.
+exist or would be inconvenient.  Using submodules is preferred over copying
+copies from other projects, however.  When a fork is needed, see the
```
Comment (CodeMonkeyLeet) [inline docs/Community/Contributing.md:320]:
```suggestions
source files from other projects, however.  When a fork is needed, see the
```

File: docs/Community/Repositories.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+necessary to use a temporary fork for some period of time, whether for
+development, or for use while the upstream submission is being reviewed
+and iterated on.  In such cases, a fork might be done under an individual's
+github organization or the openenenclave organization.  (A fork is not
+needed at all if a branch can be created in the related project.)
+
+If the Open Enclave SDK needs to directly depend on a fork of the related
+project, it is generally better for the fork to be done under the openenclave
+organization.  This makes it clear that Open Enclave itself, rather than
+a just a random developer, depends on the fork.
```
Comment (CodeMonkeyLeet) [inline docs/Community/Repositories.md:58]: Redundant "~~a~~ just a" but I'd reword it to not use "random developer" as it sounds pejorative.
```suggestion
organization.  This makes it clear that fork is a dependency for the Open Enclave
SDK and is not intended for general public use.
```

File: docs/Community/Repositories.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+   1. Must adopt the Open Enclave [Code of Conduct](Governance.md#code-of-conduct)
+      statement in their repo.
+   2. Must use the same license and copyright rules as the forked project.
+   3. Must adopt the [DCO](Contributing.md#developer-certificate-of-origin)
+      bot automation for pull requests
+      unless otherwise approved by the [CGC](governance/README.md).
+   4. Must follow other [contributing guidelines](Contributing.md)
+      unless otherwise approved by the [CGC](governance/README.md).
+   5. Must live under `github.com/openenclave/<project-name>`.
+   6. Repository must be approved by [SIG-Architecture](sig-architecture/README.md).
```
Comment (CodeMonkeyLeet) [inline docs/Community/Repositories.md:71]:
```suggestion
The repository must:
   1. Adopt the Open Enclave [Code of Conduct](Governance.md#code-of-conduct)
      statement in their repo.
   2. Use the same license and copyright rules as the forked project.
   3. Adopt the [DCO](Contributing.md#developer-certificate-of-origin)
      bot automation for pull requests
      unless otherwise approved by the [CGC](governance/README.md).
   4. Follow the other [contributing guidelines](Contributing.md)
      unless otherwise approved by the [CGC](governance/README.md).
   5. Be created under `github.com/openenclave/<project-name>`.
   6. Be approved by [SIG-Architecture](sig-architecture/README.md).
```

File: docs/Community/Repositories.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+[Open Enclave issue](https://github.com/openenclave/openenclave/issues)
+if one does not already exist.
+
+In addition to stating any deviations from the Open Enclave defaults
+discussed under Rules above, any request to create a new repository
+should also answer the following questions:
+
+1. Should github issues be filed in the new repo, or in the main
+   openenclave repo?
+2. Should issues in the new repo be discussed in the master openenclave
```
Comment (CodeMonkeyLeet) [inline docs/Community/Repositories.md:87]: This can probably be lumped in with 1 under the broader question of how issues with the new repo should be tracked and triaged.

I think a more salient question that must always be addressed is who is responsible for maintaining the new repo, how that contact info is reflected, and how "inactivity" should be measured for a given repo (particularly given the removal criteria).

File: docs/Community/Repositories.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+   triage meeting, or in some separate process?
+3. Should one or more of the existing SIGs (e.g.,
+   [SIG-Architecture](sig-architecture/README.md),
+   [SIG-Testing](sig-testing/README.md)) oversee relevant aspects of the
+   new repo, or not?
+4. Will the repo be public or private, and if private, why?
```
Comment (CodeMonkeyLeet) [inline docs/Community/Repositories.md:93]: I would suggest that this document only deals with public repos since we don't really discuss what might be reasonable criteria for asking for a private repo to be created. 

The one standard use case that we have for that is security advisories, and that will be addressed specifically under the documentation for that process.

File: docs/Community/Repositories.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+### Grounds for removal
+
+Core repositories may be removed from the project if they
+are deemed _inactive_. Inactive repositories are those that meet any of the
+following criteria:
+
+   * There are no longer any active maintainers for the project and no
+     replacements can be found.
+   * All PRs or Issues have gone un-addressed for longer than six months.
+   * There have been no new commits or other changes in more than a year.
```
Comment (CodeMonkeyLeet) [inline docs/Community/Repositories.md:121]: Is it unreasonable to allow for some repos that may not need to churn on a regular basis (e.g. samples against stable OE SDK APIs)?

Comment (dthaler) [reply docs/Community/Repositories.md:121]: This bullet was copied from the Kubernetes one, but I'll remove it, since I agree with the point and line 120 already covers the case where no changes have been done when some are actually needed.

File: docs/Community/Contributing.md (hunk 1 above)
Comment (dthaler) [reply docs/Community/Contributing.md]: Done

File: docs/Community/Repositories.md (hunk 2 above)
Comment (dthaler) [reply docs/Community/Repositories.md]: Done

File: docs/Community/Repositories.md (hunk 3 above)
Comment (dthaler) [reply docs/Community/Repositories.md]: Done

File: docs/Community/Repositories.md (hunk 4 above)
Comment (dthaler) [reply docs/Community/Repositories.md]: Done, except I didn't mention inactivity here since the removal section comes later.

File: docs/Community/Repositories.md (hunk 5 above)
Comment (dthaler) [reply docs/Community/Repositories.md]: Done

**Review by CodeMonkeyLeet — APPROVED** [review]
LGTM

**Review by achamayou — APPROVED** [review]

**Comment by anakrish** [comment]
@BRMcLaren Can you weigh in from CI perspective?

**Comment by BRMcLaren** [comment]
@anakrish  Thanks for reminding me.

@dthaler Something that I see in the Kubernetes project is rules around [donating](https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md#rules-for-donated-repositories). Is this something we can add or bundle in with "Creating" repositories?

The other thing to mention is that there is a decoupling effort for the CI infrastructure to resemble the Kubernetes [efforts](https://github.com/kubernetes/test-infra ) and it will raise questions regarding permissions for new repos, onboarding, and adding bots and if they are required vs optional. These are covered by the kubernetes project by these rules.

1. Must adopt all Kubernetes automation (e.g. /lgtm, etc)
2. All OWNERS must be members of standing as defined by ability to vote in Kubernetes steering committee elections. in the Kubernetes community

Something that is coming is our adoption (assuming approved) of the same automation and it makes me wonder if these rules will be easily malleable in the future or if it will be hard to retroactively implement.

**Review by radhikaj — APPROVED** [review]

**Comment by dthaler** [comment]
> Something that I see in the Kubernetes project is rules around [donating](https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md#rules-for-donated-repositories). Is this something we can add or bundle in with "Creating" repositories?

@BRMcLaren  Yes I consider it already part of "Creating" repositories.  The rules are the same and I didn't see a need to mention all the possible reasons why you might be creating a repository, so didn't call it (or any of the many other cases) out explicitly.

> The other thing to mention is that there is a decoupling effort for the CI infrastructure to resemble the Kubernetes [efforts](https://github.com/kubernetes/test-infra) and it will raise questions regarding permissions for new repos, onboarding, and adding bots and if they are required vs optional. These are covered by the kubernetes project by these rules.
> 
> 1. Must adopt all Kubernetes automation (e.g. /lgtm, etc)
> 2. All OWNERS must be members of standing as defined by ability to vote in Kubernetes steering committee elections. in the Kubernetes community
> 
> Something that is coming is our adoption (assuming approved) of the same automation and it makes me wonder if these rules will be easily malleable in the future or if it will be hard to retroactively implement.

I didn't follow whether you wanted some specific wording somewhere or just something to think about outside the PR.   If you have a specific suggestion, feel free to suggest text.  Otherwise would you be ok with merging this and handling it separately?

**Comment by BRMcLaren** [comment]
@dthaler Thanks, sure sounds good.

File: docs/Community/Repositories.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+1. How should issues with the new repo be tracked and triaged?
+   Should github issues be filed in the new repo, or in the main
+   openenclave repo?  Should they be discussed in the master openenclave
+   triage meeting, or in some separate process?
+2. Should one or more of the existing SIGs (e.g.,
+   [SIG-Architecture](sig-architecture/README.md),
+   [SIG-Testing](sig-testing/README.md)) oversee relevant aspects of the
+   new repo, or not?
+3. Who is reponsible for maintaining the new repo?  How is their
```
Comment (AevaOnline) [inline docs/Community/Repositories.md:95]: I suggest the recording format for this 3rd item be standardized, so that the maintainer-list be more easily integrated with any relevant project-wide automation.

Comment (BRMcLaren) [reply docs/Community/Repositories.md:95]: The way the [oeedger8r](https://github.com/openenclave/oeedger8r-cpp/blob/master/OWNERS) project handles this regarding the "who" I believe is ideal and it ties into automation easily.

**Comment by AevaOnline** [comment]
One minor suggestion aside, I think this is a good change.

/lgtm

File: docs/Community/Repositories.md (hunk 7 above)
Comment (BRMcLaren) [reply docs/Community/Repositories.md:95]: Opening #3640 for standardization of the third.

**Review by johnkord — APPROVED** [review]

**Review by jhand2 — APPROVED** [review]

**Review by AevaOnline — APPROVED** [review]

**Review by anakrish — APPROVED** [review]
LGTM

**Comment by dthaler** [comment]
bors try

**Comment by dthaler** [comment]
> 
>  > git fetch --tags --progress -- https://github.com/openenclave/openenclave +refs/heads/*:refs/remotes/origin/* # timeout=10
> 
> ERROR: Error cloning remote repo 'origin'
> 
> hudson.plugins.git.GitException: Command "git fetch --tags --progress -- https://github.com/openenclave/openenclave +refs/heads/*:refs/remotes/origin/*" returned status code 128:
> 
> stdout: 
> 
> stderr: fatal: unable to access 'https://github.com/openenclave/openenclave/': Failed to connect to github.com port 443: Timed out

**Comment by dthaler** [comment]
bors retry
